go/internal/feast/sqliteonlinestore.go

```diff
@@ -0,0 +1,159 @@
+	// Feast project name
+	project string
+	path    string
+	db      *sql.DB
+}
+
+// Creates a new sqlite online store object. onlineStoreConfig should have relative path of database file with respect to repoConfig.repoPath.
+func NewSqliteOnlineStore(project string, repoConfig *RepoConfig, onlineStoreConfig map[string]interface{}) (*SqliteOnlineStore, error) {
+	store := SqliteOnlineStore{project: project}
+	if db_path, ok := onlineStoreConfig["path"]; !ok {
+		return nil, fmt.Errorf("cannot find sqlite path %s", db_path)
+	} else if dbPathStr, ok := db_path.(string); !ok {
+		return nil, fmt.Errorf("cannot find convert sqlite path to string %s", db_path)
+	} else {
+		store.path = fmt.Sprintf("%s/%s", repoConfig.RepoPath, dbPathStr)
+		db, err := initializeConnection(store.path)
+		if err != nil {
+			return nil, err
+		}
+		store.db = db
+	}
+	return &store, nil
+}
+
+func (s *SqliteOnlineStore) Destruct() {
+	s.db.Close()
+}
+
+// Returns FeatureData 2D array. Each row corresponds to one entity value and each column corresponds to a single feature where the number of columns should be
+// same length as the length of featureNames. Reads from every table in featureViewNames with the entity keys described.
+func (s *SqliteOnlineStore) OnlineRead(ctx context.Context, entityKeys []types.EntityKey, featureViewNames []string, featureNames []string) ([][]FeatureData, error) {
+	featureCount := len(featureNames)
+	_, err := s.getConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	project := s.project
+	results := make([][]FeatureData, len(entityKeys))
+	entityNameToEntityIndex := make(map[string]int)
+	in_query := make([]string, len(entityKeys))
+	serialized_entities := make([]interface{}, len(entityKeys))
+	for i := 0; i < len(entityKeys); i++ {
+		serKey, err := serializeEntityKey(&entityKeys[i])
+		if err != nil {
+			return nil, err
+		}
+		// TODO: fix this, string conversion is not safe
+		entityNameToEntityIndex[string(*serKey)] = i
+		// for IN clause in read query
+		in_query[i] = "?"
+		serialized_entities[i] = *serKey
+	}
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	for idx := 0; idx < len(entityKeys); idx++ {
+		results[idx] = make([]FeatureData, featureCount)
+	}
+	for _, featureViewName := range featureViewNames {
+		query_string := fmt.Sprintf(`SELECT entity_key, feature_name, value, event_ts
+									FROM %s
+									WHERE entity_key IN (%s)
+									ORDER BY entity_key`, tableId(project, featureViewName), strings.Join(in_query, ","))
+		rows, err := s.db.Query(query_string, serialized_entities...)
+		if err != nil {
+			return nil, err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var entity_key string
+			var feature_name string
+			var valueString string
+			var event_ts time.Time
+			var value types.Value
+			err = rows.Scan(&entity_key, &feature_name, &valueString, &event_ts)
+			if err != nil {
+				log.Fatal(err)
```

Comment:
Why are something logged and other things returned as `fmt.Errorf`? Also a `log.Fatal` will crash the running process which is something we should definietly not do in this code path - that should be reserved for unrecoverable errors like invalid config at start up time or something.

Reply:
This was part of my testing code and accidentally got left in there, thanks for the catch.